src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+
+        // TODO: Verify that this RecordBatch contains partition values that agree with previous partition values (if any) for the current writer context
+        self.partition_values = partition_values;
+
+        // TODO: collect/update column stats
+
+        // write the record batch to the held arrow writer
+        self.arrow_writer.write(record_batch)?;
+
+        // increment num records (for the entire file) by the number of records in the handled batch
+        self.num_records += record_batch.num_rows() as i64;
+
+        Ok(())
+    }
+
+    /// Returns the current byte length of the in memory buffer.
+    /// This should be used by the caller to decide when to call `complete_file`.
+    pub fn buffer_len(&self) -> Result<usize, DeltaWriterError> {
+        Ok(self.cursor.data().len())
+    }
+    
+    /// Writes the existing parquet bytes to storage and resets internal state to handle another round.
+    /// Returns an `Add` action representing the added file to be appended to the delta log.
+    pub async fn complete_file(&mut self, partition_cols: &Vec<String>) -> Result<Add, DeltaWriterError> {
+        let path = self.next_data_path(partition_cols, &self.partition_values).unwrap();
+
+        let obj_bytes = self.cursor.data();
+        let file_size = obj_bytes.len() as i64;
+
+        let storage_path = format!("{}/{}", self.table_path, path);
+
+        self.storage.put_obj(storage_path.as_str(), obj_bytes.as_slice()).await?;
+
+        // Close the arrow writer to flush remaining bytes and write the parquet footer
+        self.arrow_writer.close()?;
+
+        let num_records = self.num_records;
+ 
+        // Re-initialize state to handle another file
+        self.reset();
+
+        // Create and return the "add" action associated with the completed file
+        let add = create_add(&self.partition_values, path, file_size, num_records);
+
+        add
+    }
+
+    fn reset(&mut self) {
+        self.cursor = InMemoryWriteableCursor::default();
+        // TODO: Handle error
+        self.arrow_writer = ArrowWriter::try_new(self.cursor.clone(), self.arrow_schema_ref.clone(), Some(self.writer_properties.clone())).unwrap();
+        self.num_records = 0;
+    }
+
+    // TODO: parquet files have a 5 digit zero-padded prefix and a "c\d{3}" suffix that I have not been able to find documentation for yet.
+    fn next_data_path(&self, partition_cols: &Vec<String>, partition_values: &HashMap<String, String>) -> Result<String, DeltaWriterError> {
+        // TODO: what does 00000 mean?
+        let first_part = "00000";
+        let uuid_part = Uuid::new_v4();
+        // TODO: what does c000 mean?
+        let last_part = "c000";
+
+        let file_name = format!("part-{}-{}-{}.parquet", first_part, uuid_part, last_part);
+
+        let data_path = if partition_cols.len() > 0 {
+            let mut path_part = String::with_capacity(20);
+
+            // ugly string builder hack
+            let mut first = true;
```

Comment:
i think you can put each path component into a vec, then join them with the path separate at the end: https://doc.rust-lang.org/std/primitive.slice.html#method.join

Reply:
Nice suggestion. Fixed in https://github.com/delta-io/kafka-delta-ingest/pull/1/commits/c90ff3418c7bbf0e56cce7a7dda990e3c616140a#diff-5d8c9545386ed0b9dcf9dab1722a6b484e1d3014d9a2470788dd7b20509ca642R176.